File: src/main/java/tn/esprit/models/Billet.java
package tn.esprit.models;

import java.time.LocalDateTime;


public class Billet {

    public enum TypeBillet {
        SIMPLE, DUO, VIP;
    }
    private int idBillet, prix;
    private String proprietaire;
    private LocalDateTime dateAchat; // Utilisation de LocalDateTime au lieu de Date
    private TypeBillet type;
    private Event event;

    // Constructeur par défaut
    public Billet() {
        // Initialisation de la date d'achat à la date actuelle
        this.dateAchat = LocalDateTime.now();
    }

    // Constructeur avec paramètres
    public Billet(int idBillet, String proprietaire, int prix, LocalDateTime dateAchat, TypeBillet type, Event event) {
        this.idBillet = idBillet;
        this.proprietaire = proprietaire;
        this.prix = prix;
        this.dateAchat = dateAchat;
        this.type = type;
        this.event = event;
    }

    public Billet(int prix, Event event, String proprietaire, TypeBillet type) {
        this.prix = prix;
        this.event = event;
        this.proprietaire = proprietaire;
        this.type = type;
        this.dateAchat = LocalDateTime.now(); // Initialisation de la date à la création de l'objet
    }

    // Getters et Setters
    public int getIdBillet() {
        return idBillet;
    }

    public void setIdBillet(int idBillet) {
        this.idBillet = idBillet;
    }

    public String getProprietaire() {
        return proprietaire;
    }

    public void setProprietaire(String proprietaire) {
        this.proprietaire = proprietaire;
    }

    public int getPrix() {
        return prix;
    }

    public void setPrix(int prix) {
        this.prix = prix;
    }

    public LocalDateTime getDateAchat() {
        return dateAchat;
    }

    public void setDateAchat(LocalDateTime dateAchat) {
        this.dateAchat = dateAchat;
    }

    public TypeBillet getType() {
        return type;
    }

    public void setType(TypeBillet type) {
        this.type = type;
    }

    public Event getEvent() {
        return event;
    }

    public void setEvent(Event event) {
        this.event = event;
    }

    // Méthode toString pour affichage
    @Override
    public String toString() {
        return "Billet{" +
                "idBillet=" + idBillet +
                ", proprietaire='" + proprietaire + '\'' +
                ", prix=" + prix +
                ", dateAchat=" + dateAchat +
                ", type=" + type +
                ", event=" + event +
                '}';
    }
}

File: src/Entity/Billet.php
<?php

namespace App\Entity;

use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Symfony\Component\Validator\Constraints as Assert;

use App\Repository\BilletRepository;

#[ORM\Entity(repositoryClass: BilletRepository::class)]
#[ORM\Table(name: 'billet')]
class Billet
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idBillet',type: 'integer')]
    private ?int $idBillet = null;

    public function getIdBillet(): ?int
    {
        return $this->idBillet;
    }

    public function setIdBillet(int $idBillet): self
    {
        $this->idBillet = $idBillet;
        return $this;
    }

    #[ORM\Column(name: 'proprietaire',type: 'string', nullable: false)]
    #[Assert\NotBlank(message: 'Le nom du propriétaire est obligatoire.')]
    #[Assert\Length(min: 3, max: 50, minMessage: 'Le nom du propriétaire doit contenir au moins {{ limit }} caractères.')]
    private ?string $proprietaire = null;

    public function getProprietaire(): ?string
    {
        return $this->proprietaire;
    }

    public function setProprietaire(string $proprietaire): self
    {
        $this->proprietaire = $proprietaire;
        return $this;
    }

    #[ORM\Column(name: 'prix',type: 'integer', nullable: false)]
    #[Assert\NotBlank(message: 'Le prix est obligatoire.')]
    #[Assert\Positive(message: 'Le prix doit être un entier positif.')]
    private ?int $prix = null;

    public function getPrix(): ?int
    {
        return $this->prix;
    }

    public function setPrix(int $prix): self
    {
        $this->prix = $prix;
        return $this;
    }

    #[ORM\Column(name: 'dateAchat',type: 'datetime', nullable: false)]
    private ?\DateTimeInterface $dateAchat = null;

    public function getDateAchat(): ?\DateTimeInterface
    {
        return $this->dateAchat;
    }

    public function setDateAchat(\DateTimeInterface $dateAchat): self
    {
        $this->dateAchat = $dateAchat;
        return $this;
    }

    #[ORM\Column(name: 'type', type: 'string', nullable: false)]
    #[Assert\NotBlank(message: 'Le type de billet est obligatoire.')]
    #[Assert\Choice(choices: ['SIMPLE', 'DUO', 'VIP'], message: 'Choisissez un type valide (SIMPLE, DUO, VIP).')]
    private ?string $type = null;

    public function getType(): ?string
    {
        return $this->type;
    }

    public function setType(string $type): self
    {
        $this->type = $type;
        return $this;
    }

    #[ORM\ManyToOne(targetEntity: Event::class, inversedBy: 'billets')]
    #[ORM\JoinColumn(name: 'idEvent', referencedColumnName: 'idEvent')]
    #[Assert\NotNull(message: 'Veuillez sélectionner un événement.')]
    private ?Event $event = null;

    public function getEvent(): ?Event
    {
        return $this->event;
    }

    public function setEvent(?Event $event): self
    {
        $this->event = $event;
        return $this;
    }

    #[ORM\OneToMany(targetEntity: Paiement::class, mappedBy: 'billet')]
    private Collection $paiements;

    public function __construct()
    {
        $this->paiements = new ArrayCollection();
    }

    /**
     * @return Collection<int, Paiement>
     */
    public function getPaiements(): Collection
    {
        if (!$this->paiements instanceof Collection) {
            $this->paiements = new ArrayCollection();
        }
        return $this->paiements;
    }

    public function addPaiement(Paiement $paiement): self
    {
        if (!$this->getPaiements()->contains($paiement)) {
            $this->getPaiements()->add($paiement);
        }
        return $this;
    }

    public function removePaiement(Paiement $paiement): self
    {
        $this->getPaiements()->removeElement($paiement);
        return $this;
    }

}
